lib/reduce_helper.cu

```diff
@@ -0,0 +1,115 @@
+#include <quda_internal.h>
+#include <quda_cuda_api.h>
+#include <malloc_quda.h>
+#include <reduce_helper.h>
+
+// These are used for reduction kernels
+static device_reduce_t *d_reduce = nullptr;
+static device_reduce_t *h_reduce = nullptr;
+static device_reduce_t *hd_reduce = nullptr;
+
+static count_t *reduce_count = nullptr;
+static cudaEvent_t reduceEnd;
+
+namespace quda
+{
+
+  namespace reducer
+  {
+
+    // FIXME need to dynamically resize these
+    void *get_device_buffer() { return d_reduce; }
+    void *get_mapped_buffer() { return hd_reduce; }
+    void *get_host_buffer() { return h_reduce; }
+    count_t *get_count() { return reduce_count; }
+    cudaEvent_t &get_event() { return reduceEnd; }
+
+    size_t buffer_size()
+    {
+      /* we have these different reductions to cater for:
+
+         - regular reductions (reduce_quda.cu) where are reducing to a
+           single vector type (max length 4 presently), with possibly
+           parity dimension, and a grid-stride loop with max number of
+           blocks = 2 x SM count
+
+         - multi-reductions where we are reducing to a matrix of size
+           of size QUDA_MAX_MULTI_REDUCE of vectors (max length 4), with
+           possible parity dimension, and a grid-stride loop with
+           maximum number of blocks = 2 x SM count
+      */
+
+      int reduce_size = 4 * sizeof(device_reduce_t);
+      int max_reduce = 2 * reduce_size;
+      int max_multi_reduce = 2 * QUDA_MAX_MULTI_REDUCE * reduce_size;
+      int max_reduce_blocks = 2 * deviceProp.multiProcessorCount;
+
+      // reduction buffer size
+      size_t bytes = max_reduce_blocks * std::max(max_reduce, max_multi_reduce);
+      return bytes;
+    }
+
+    // need to placement new constructor to initialize the atomic counters
+    template <typename T> __global__ void init_count(T *counter, int n)
+    {
+      for (int i = 0; i < n; i++) new (counter + i) T {0};
+    }
+
+    void init()
+    {
+#ifdef HETEROGENEOUS_ATOMIC
+      // fast reductions use heterogeneous std::atomic to synchronize CPU and GPU as opposed to using CUDA events
+      warningQuda("Experimental heterogeneous-atomic reductions enabled");
```

Comment:
guess that should go away when we no longer consider it experimental

Reply:
Yes, I should just delete that.  Will do so in next push.